### PR TITLE
feat: Handle symlinks, support pnpm and monorepo,

### PR DESCRIPTION
**Pull request by rxliuli**
ref: https://github.com/develar/app-builder/issues/84

### CODE REVIEW

**Comment by rxliuli** [comment]
@develar Any progress?

File: pkg/fs/dirname.go [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+import (
+	"path"
+	"runtime"
+)
+
+func Dirname() string {
+	_, filename, _, _ := runtime.Caller(1)
+	return path.Dir(filename)
+}
```
Comment (BlackHole1) [inline pkg/fs/dirname.go:11]:
```suggestion
import (
	"path"
	"os"
)

func exedir() {
	exe, err := os.Executable()
	if err != nil {
		panic(err)
	}
	
	return path.Dir(exe)
}

var Exedir = exedir()
```

Create and move to `pkg/test/helper_test.go`

File: pkg/fs/findParent.go [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+func FindParent(cwd string, predicate Predicate) string {
+	if predicate(cwd) {
+		return cwd
+	}
+	parent := filepath.Dir(cwd)
+	if parent == cwd {
+		return ""
+	}
+	return FindParent(parent, predicate)
+}
```
Comment (BlackHole1) [inline pkg/fs/findParent.go:20]:
```suggestion
func FindParentWithFile(cwd string, file string) string {
	if fs.PathExists(path.Join(cwd, file)) {
		return cwd
	}
	parent := filepath.Dir(cwd)
	if parent == cwd {
		return ""
	}
	return FindParent(parent, file)
}
```

File: pkg/fs/pathExists.go [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+func PathExists(path string) bool {
+	_, err := os.Stat(path)
+	if err == nil {
+		return true
+	}
+	if os.IsNotExist(err) {
+		return false
+	}
+	return false
+}
```
Comment (BlackHole1) [inline pkg/fs/pathExists.go:16]:
```suggestion
func PathExists(path string) bool {
	_, err := os.Stat(path)
	if err != nil {
		return false
	}

	return true
}
```

File: pkg/node-modules/cli_test.go [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+	Version string `json:"version"`
+}
+
+type NodeTreeItem struct {
+	Dir  string            `json:"dir"`
+	Deps []NodeTreeDepItem `json:"deps"`
+}
+
+func nodeDepTree(t *testing.T, dir string) {
+
```
Comment (BlackHole1) [inline pkg/node-modules/cli_test.go:26]:
```suggestion
```

**Review by BlackHole1 — CHANGES_REQUESTED** [review]

File: pkg/fs/dirname.go (hunk 1 above)
Comment (rxliuli) [reply pkg/fs/dirname.go:11]: The paths obtained by runtime.Caller and os.Executable are not equivalent. What is expected in the test is to obtain the path to the test source code, so the implementation will not be modified here

ref

```go
_, Caller, _, _ := runtime.Caller(0)
fmt.Println("Caller", Caller) // app-builder/pkg/node-modules/cli_test.go

exe, _ := os.Executable()
fmt.Println("Executable", exe) //~/AppData\Local\Temp\go-build2984811955\b001\node-modules.test.exe
```

File: pkg/fs/findParent.go (hunk 2 above)
Comment (rxliuli) [reply pkg/fs/findParent.go:20]: Yes, this can be simplified to a more single-function function

File: pkg/node-modules/cli_test.go (hunk 4 above)
Comment (rxliuli) [reply pkg/node-modules/cli_test.go:26]: Why should this test be deleted?

File: pkg/node-modules/cli_test.go [hunk 5]
```diff
@@ -50,6 +50,14 @@ func nodeDepTree(t *testing.T, dir string) {
...
+	nodeDepTree(t, path.Join(Dirname(), "npm-demo"))
+	nodeDepTree(t, path.Join(Dirname(), "pnpm-demo"))
+}
+
+func TestDirname(t *testing.T) {
+	_, Caller, _, _ := runtime.Caller(0)
+	fmt.Println("Caller", Caller)
+
+	exe, _ := os.Executable()
+	fmt.Println("Executable", exe)
```
Comment (BlackHole1) [inline pkg/node-modules/cli_test.go:62]:
```suggestion
```

File: pkg/fs/pathExists.go [hunk 6]
```diff
@@ -6,11 +6,5 @@ import (
 
 func PathExists(path string) bool {
```
Comment (BlackHole1) [inline pkg/fs/pathExists.go:7]: Move to `findParent.go` file. and rename to: `pathExists`

File: pkg/fs/findParent.go [hunk 7]
```diff
@@ -1,20 +1,20 @@
 package fs
 
 import (
+	"path"
 	"path/filepath"
 )
 
 // Predicate is a function that takes a directory and returns a boolean.
 type Predicate func(dir string) bool
```
Comment (BlackHole1) [inline pkg/fs/findParent.go]: remove

**Review by BlackHole1 — CHANGES_REQUESTED** [review]

File: pkg/node-modules/cli_test.go (hunk 4 above)
Comment (BlackHole1) [reply pkg/node-modules/cli_test.go]: My meaning is "delete the empty line.😃

File: pkg/fs/dirname.go (hunk 1 above)
Comment (BlackHole1) [reply pkg/fs/dirname.go]: Perhaps you need to add `// +build:ignore` in file header

**Comment by rxliuli** [comment]
@BlackHole1 Is there anything blocking it now?

**Comment by psd-coder** [comment]
Hey guys! Any progress with this PR?

**Comment by rxliuli** [comment]
@BlackHole1 any progress?

**Comment by rxliuli** [comment]
@BlackHole1 any progress?

**Comment by BlackHole1** [comment]
Hi @rxliuli. I don’t have the authority to merge this PR, I can only assist with code review. I tried contacting xx before, but it was unsuccessful.

**Comment by miaulightouch** [comment]
I'm glad to see that this pull request has been merged. I hope there will be a new release soon.

**Comment by mmaietta** [comment]
Are tests working for anyone? I run `app-builder/pkg/node-modules> go test` and I receive
```
--- FAIL: TestNodeDepTreeCmd (0.61s)
    cli_test.go:41: 
        Expected
            <[]string | len:0, cap:0>: []
        to consist of
            <[]interface {} | len:1, cap:1>: [
                ["react", "js-tokens", "loose-envify"],
            ]
--- FAIL: TestReadDependencyTreeByNpm (0.00s)
    nodeModuleCollector_test.go:31: 
        Expected
            <[]string | len:0, cap:0>: []
        to consist of
            <[]interface {} | len:1, cap:1>: [
                ["js-tokens", "react", "loose-envify"],
            ]
--- FAIL: TestReadDependencyTreeByPnpm (0.00s)
    nodeModuleCollector_test.go:56: 
        Expected
            <[]string | len:0, cap:0>: []
        to consist of
            <[]interface {} | len:1, cap:1>: [
                ["js-tokens", "react", "loose-envify"],
            ]
FAIL
exit status 1
FAIL	github.com/develar/app-builder/pkg/node-modules	0.618s

```

**Pull request merged**